fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -551,82 +515,68 @@ def _setup_bucket_strategy(self) -> None:
             dst_rank = self._trainable_param_to_rank[param]
 
             if param.device not in self.buckets.keys():
-                self.buckets[param.device] = [
-                    Bucket(buffer=torch.zeros(self.buffer_max_size, dtype=param.dtype, device=device))
-                    for _ in range(dist.get_world_size(self.process_group))
-                ]
-
-            bucket = self.buckets[device][dst_rank]
-            bucket.destination = self._local_to_global_rank[dst_rank]
+                self.buckets[param.device] = {}
+
+            if dst_rank not in self.buckets[device].keys():
+                self.buckets[param.device][dst_rank] = GradBucket(
+                    size=self.buffer_max_size,
+                    dtype=param.dtype,
+                    device=device,
+                    destination=self._local_to_global_rank[dst_rank],
+                    process_group=self.process_group,
+                )
 
             # Criteria to decide whether this parameter is to be bucketed or not:
             # - enough room in the bucket
-            if (bucket.fill + param.numel()) < self.buffer_max_size:
-                self._should_bucket_grad.append(True)
+            if self.buckets[device][dst_rank].can_add_grad_view(param):
+                self._should_bucket_grad[index] = True
+                self.buckets[device][dst_rank].add_grad(param)
 
-                # This parameter gradients becomes a view of the bucket
-                fill_next = bucket.fill + param.numel()
+        self._bucket_list = list(chain(*[self.buckets[device].values() for device in self.buckets.keys()]))
 
-                if param.grad is None:
-                    # will be overwritten just below, see next line
-                    param.grad = torch.zeros_like(param)
-
-                param.grad.data = bucket.buffer[bucket.fill : fill_next].view_as(param.data)
-                bucket.fill = fill_next
-
-                # Update the bucket
-                self.buckets[device][dst_rank].max_params_checked_in += 1
+        # Post fix:
+        for bucket in self._bucket_list:
+            # - resize the buckets to remove lost space in the end
+            bucket.shrink()
 
-            else:
-                self._should_bucket_grad.append(False)
+            # - store the callback which releases the bucket
+            def cleanup(fut: torch.futures.Future) -> None:
+                if bucket.destination != self.global_rank:
+                    fut.wait()
+                    # FIXME: Autograd dies with this
+                    # bucket.collapse()
 
-        self._bucket_list = list(chain(*[self.buckets[device] for device in self.buckets.keys()]))
+            bucket.callback = cleanup
 
-        # Resize the buckets to remove lost space in the end
-        for bucket in self._bucket_list:
-            bucket.buffer.resize_(bucket.fill)
-            bucket.sent = True
-
-    def _consume_work_handles(self) -> None:
```

Comment:
all this was nice, but handle.wait() does not do the same thing as (torch.future).wait()... this was an expensive mistake